**Introduction**

Linear Algebra is about working on linear systems of equations (linear regression is an example: y = AX).
https://machinelearningmastery.com/linear-algebra-machine-learning/

* Linear algebra is the most useful mathematics we have learnt.

* Linear algebra is what machine learning is doing under the hood.

**Real applications of Linear Algebra**

* 1) Dataset - which is a basically a matrix 
* 2) Image processing - Pixel in image
* 3) Natural Language Processing - Tokenization, lemmanisation and normalization involves matrix
* 4) Dimensionality reduction
* 5) Deep Learning

https://www.quora.com/What-are-the-applications-of-linear-algebra-in-machine-learning

Top packages used for AI projects include:

1) **Numpy** - a package with support for mathematical operations on multidimensional data is heavily used with more than 75% of AI projects on GitHub are using it.

2) **Scipy** - a scientific computation toolkit

https://venturebeat.com/2019/01/24/github-numpy-and-scipy-are-the-most-popular-packages-for-machine-learning-projects/

Linear algebra routines and matrix decompositions are provided in both numpy.linalg and scipy.linalg.

scipy provides Linear algebra routines and matrix decompositions extending beyond those provided in numpy.

Linear algebra, such as matrix multiplication, decompositions, determinants, and other square matrix math, is an important part of any array library. 

Multiplying two two-dimensional arrays with * is an element-wise product instead of a matrix dot product. 

There is a function dot, both an array method and a function in the numpy namespace, for matrix multiplication:

In [1]:
import numpy as np

x = np.array([[4., 1., 3.], [4., 1., 6.]])
y = np.array([[2., 21.], [-1, 7], [8, 9]])
z = np.ones(3)

print("x")
print(x)

print("y")
print(y)

print("z")
print(z)

x
[[4. 1. 3.]
 [4. 1. 6.]]
y
[[ 2. 21.]
 [-1.  7.]
 [ 8.  9.]]
z
[1. 1. 1.]


x.dot(y) is equivalent to np.dot(x, y)

In [2]:
print("x.dot(y)")
print(x.dot(y))

x.dot(y)
[[ 31. 118.]
 [ 55. 145.]]


In [3]:
print("x.dot(z)")
print(x.dot(z))

## This is equivalent to np.dot(x,y)
print("np.dot(x,y)")
print(np.dot(x,y))

print("np.dot(x,z)")
print(np.dot(x,z))

x.dot(z)
[ 8. 11.]
np.dot(x,y)
[[ 31. 118.]
 [ 55. 145.]]
np.dot(x,z)
[ 8. 11.]


In [4]:
## This is equivalent to x @ y
print("x @ y")
print(x @ y)

print("x @ z")
print(x @ z)

x @ y
[[ 31. 118.]
 [ 55. 145.]]
x @ z
[ 8. 11.]


The @ symbol (as of Python 3.5) also works as an infix operator that performs matrix multiplication:


numpy.linalg has a standard set of matrix decompositions and things like inverse and determinant. These are implemented under the hood via the same industrystandard linear algebra libraries.

Most commonly used functions of numpy.linalg
1. diag - returns the diagonal (or off-diagonal) elements of a square matrix as a 1D array, or convert a 1D array into a square matrix with zeros on the off-diagonal
2. dot - Performs Matrix multiplication
3. trace - Computes the sum of the diagonal elements
4. det - Computes the matrix determinant
5. qr - Factors the matrix a as qr, where q is orthonormal and r is upper-triangular.

In [5]:
print("x")
print(x)
print("\nDiagonal of the matrix x")
print(np.diag(x))

x
[[4. 1. 3.]
 [4. 1. 6.]]

Diagonal of the matrix x
[4. 1.]


In [6]:
X = np.array([[4., 1., 5.], [1., 0., 1], [1., 0., 0.]])
print("X")
print(X)
print("\nDeterminant of the matrix X")
print(np.linalg.det(X))

X
[[4. 1. 5.]
 [1. 0. 1.]
 [1. 0. 0.]]

Determinant of the matrix X
1.0


In [7]:
print("X")
print(X)
print("\nInverse of the matrix X")
XInv = np.linalg.inv(X)
print(XInv)

X
[[4. 1. 5.]
 [1. 0. 1.]
 [1. 0. 0.]]

Inverse of the matrix X
[[ 0.  0.  1.]
 [ 1. -5.  1.]
 [-0.  1. -1.]]


In [8]:
print("\nCheck if XInv is the inverse of X")
Z = X.dot(XInv)
print(Z)


Check if XInv is the inverse of X
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [9]:
# Computes the sum of the diagonal elements
print("X")
print(X)
print("Compute the sum of the diagonal elements of X")
print(np.trace(X))

X
[[4. 1. 5.]
 [1. 0. 1.]
 [1. 0. 0.]]
Compute the sum of the diagonal elements of X
4.0


*qr* factors the matrix a as qr, where q is orthonormal and r is upper-triangular.

In [10]:
q, r = np.linalg.qr(X)
print("X")
print(X)
print("\nq which is an orthogonal matrix decomposed from X\n")
print(q)
print("\nr which is an upper triangular matrix decomposed from X\n")
print(r)

X
[[4. 1. 5.]
 [1. 0. 1.]
 [1. 0. 0.]]

q which is an orthogonal matrix decomposed from X

[[-0.94280904  0.33333333  0.        ]
 [-0.23570226 -0.66666667 -0.70710678]
 [-0.23570226 -0.66666667  0.70710678]]

r which is an upper triangular matrix decomposed from X

[[-4.24264069 -0.94280904 -4.94974747]
 [ 0.          0.33333333  1.        ]
 [ 0.          0.         -0.70710678]]


In [11]:
print("\nTranspose of orthogonal matrix q")
print(q.T)

print("\nInverse of orthogonal matrix q")
print(np.linalg.inv(q))


Transpose of orthogonal matrix q
[[-0.94280904 -0.23570226 -0.23570226]
 [ 0.33333333 -0.66666667 -0.66666667]
 [ 0.         -0.70710678  0.70710678]]

Inverse of orthogonal matrix q
[[-9.42809042e-01 -2.35702260e-01 -2.35702260e-01]
 [ 3.33333333e-01 -6.66666667e-01 -6.66666667e-01]
 [ 5.88784672e-17 -7.07106781e-01  7.07106781e-01]]


**Note**

A matrix Q is orthogonal if its transpose is equal to its inverse.

**Linear regression **

**It is also called Ordinary Least Squares (OLS) Regression is perhaps the widely used technique in Statistical Learning to predict a continuous numerical variable. It dates back to the eighteenth century and work of Carl Friedrich Gauss and Adrien Marie Legendre. **

**Ordinary Least Squares is the simplest and most common estimator in which the coefficients are chosen to minimize the square of the distance between the predicted values and the actual values.**

**Let us solve a simple linear regression problem using both Linear algebra and the statsmodel version of OLS and check whether both produce identical results**

Statsmodels is a Python package that provides a complement to scipy for statistical computations including descriptive statistics and estimation and inference for statistical models.
It integrates well with the pandas and numpy libraries

Y = $\alpha + \beta X$
where 
α indicates the value of Y when X is equal to zero (also known as the intercept).
β indicating the slope of the line (also known as the regression coefficient). 

The relationship between variables Y and X is described using the equation of the line of best fit.
The regression coefficient β describes the change in Y that is associated with a unit change in X.  

**Let us use a simple dataset built-in Python to explain the above points.**

**About iris data set **

* Iris is a genus of 260–300 species of flowering plants with showy flowers.
* The iris flower data is a multi-variate data set collected to quantify the morphologic variation of Iris flowers of three related species, namely *Versicolor, Setosa and Virginica*.
* The data set contains 50 samples from each of three species of Iris mentioned above.

Four features were measured in centi-meter from each sample:
* Length of Sepals
* Length of Petals
* Width of Sepals
* Width of Petals

Note:
* Sepal is one of the separate, usually green parts extending from the base of a flower
* Petals are modified leaves that surround the reproductive parts of flowers

https://en.wikipedia.org/wiki/Sepal
https://en.wikipedia.org/wiki/Petal

https://medium.com/codebagng/basic-analysis-of-the-iris-data-set-using-python-2995618a6342

In [12]:
import pandas          as     pd
import numpy           as     np
from   numpy.linalg    import inv, qr, det
from   IPython.display import Image
from   sklearn         import datasets
import statsmodels.api as     sm

**Load iris dataset into an object, iris **

In [13]:
iris = datasets.load_iris()

List the keys of the object, iris

In [14]:
print(iris.keys())

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])


**Find out the number of observations and the columns in the data set, iris**

In [15]:
print(iris.data.shape) # Exploring first key, data

(150, 4)


**Inference **

There are 150 observations and 4 columns in the *iris* data set

**Get the list of 5 values of the target variable, species**

In [16]:
print(iris.target[:5]) # Exploring second key, target

[0 0 0 0 0]


**Get the name of the target variable**

In [17]:
print(iris.target_names) # Exploring third  key, target_names

['setosa' 'versicolor' 'virginica']


**Print the data dictionary for iris dataset **

In [18]:
print(iris.DESCR) # Exploring fourth  key, DESCR

Iris Plants Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20  0.76     0.9565  (high!)
    ============== ==== ==== ======= ===== ====================

    :Missing Attribute Values: None
  

Print the feature names (aka independent variable names)

In [19]:
print(iris.feature_names) # Exploring fifth key, feature_names

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


Convert iris object to a data frame 

In [20]:
iris_df  = pd.DataFrame(iris.data, columns = iris.feature_names)

**Add the target variable, species to the data frame, iris_df **

In [21]:
iris_df['Species'] = iris.target

**Print first five rows of the data frame, iris_df **

In [22]:
print(iris_df.head().T)

                     0    1    2    3    4
sepal length (cm)  5.1  4.9  4.7  4.6  5.0
sepal width (cm)   3.5  3.0  3.2  3.1  3.6
petal length (cm)  1.4  1.4  1.3  1.5  1.4
petal width (cm)   0.2  0.2  0.2  0.2  0.2
Species            0.0  0.0  0.0  0.0  0.0


In [23]:
print(iris_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
sepal length (cm)    150 non-null float64
sepal width (cm)     150 non-null float64
petal length (cm)    150 non-null float64
petal width (cm)     150 non-null float64
Species              150 non-null int32
dtypes: float64(4), int32(1)
memory usage: 5.4 KB
None


Let us try to predict Sepal length from sepal width (cm), petal length (cm), petal width (cm) of the species, Setosa.
We shall select all rows belonging to class 0, Setosa

In [24]:
setosa_df = iris_df.loc[iris_df['Species'] == 0 ]

**Define the response variable, y as the sepal length**

In [25]:
y = setosa_df[['sepal length (cm)']]

In [26]:
y

sepal length (cm)
0                 5.1
1                 4.9
2                 4.7
3                 4.6
4                 5.0
5                 5.4
6                 4.6
7                 5.0
8                 4.4
9                 4.9
10                5.4
11                4.8
12                4.8
13                4.3
14                5.8
15                5.7
16                5.4
17                5.1
18                5.7
19                5.1
20                5.4
21                5.1
22                4.6
23                5.1
24                4.8
25                5.0
26                5.0
27                5.2
28                5.2
29                4.7
30                4.8
31                5.4
32                5.2
33                5.5
34                4.9
35                5.0
36                5.5
37                4.9
38                4.4
39                5.1
40                5.0
41                4.5
42                4.4
43                5.0
44                5.1
45                4.8
46                5.1
47                4.6
48                5.3
49                5.0

**Define the predictor variables, X comprising *sepal width (cm), petal length (cm), petal width (cm) **

In [27]:
X = setosa_df[['sepal width (cm)', 'petal length (cm)', 'petal width (cm)']]

In [28]:
X = sm.add_constant(X) # To add the intercept

We are trying to form a linear regression equation, y = b0 + b1X1 + b2X2 + b3X3
* where y represents the sepal length in cm of Setosa species of iris flower
* X1 represents its petal length (cm) and X2 represents petal width (cm)
* b0 is a regression constant
* b1, b2 and b3 are regression coefficients.

Linear algebra can step in and simplify this notation

Y = Xb                                                                    

We need to solve b, which is a vector of length 50

In [29]:
print("\nHere the dimension of Y is " + str(y.shape))


Here the dimension of Y is (50, 1)


**So, Y is a vector of length 50.**

In [30]:
print("\nHere the dimension of X is " + str(X.shape))


Here the dimension of X is (50, 4)


**So, X is a matrix of  dimension 50 X 4 and X is not a square matrix and hence not invertible. **

**How do you make the non-square matrix X as a square matrix? **

By multiplying X with $X^T$ (X Transpose- rows and columns interchanged), we get a square matrix.

For example, try this

In [31]:
X1   =  np.array([[1., 7., 2.], [0., -1., 2]])
print("\nX is a matrix of shape " + str(X1.shape))


X is a matrix of shape (2, 3)


Multiple X with $X^T$

In [32]:
mat = X1.T.dot(X1)
print(mat)

[[ 1.  7.  2.]
 [ 7. 50. 12.]
 [ 2. 12.  8.]]


In [33]:
print("\nmat is a matrix of shape " + str(mat.shape))


mat is a matrix of shape (3, 3)


So, we have seen by multiplying a non-square matrix, X with its transpose we get a square matrix.

**1)  Linear algebra version of the Ordinary Least Square equation**

Let us multiply $X^T$ on both sides of the equation, Y = Xb

We get $YX^T$ = $XX^Tb$

Divide both sides of the above equation with $XX^T$ 

We get, $ {(XX^T)}^{-1} X^TY$ = $ {(XX^T)}^{-1}(XX^T)b $ 

We know that any matrix when multipled with its inverse is an identity matrix as shown below:

In [34]:
Z   =  np.array([[1., 5., 2.], [0., -1., 2], [0., 0., 1.]])

print("\nMatrix is \n%s \n"% Z)
print(det(Z))


Matrix is 
[[ 1.  5.  2.]
 [ 0. -1.  2.]
 [ 0.  0.  1.]] 

-1.0


Multiply matrix, mat by its inverse, we get an identity matrix.

In [35]:
res = Z.dot(inv(Z)) 
print(res)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


$ {(XX^T)}^{-1} X^TY$ = $ {(XX^T)}^{-1}(XX^T)b $ 

This can be re-written as $ {(XX^T)}^{-1} X^TY$ = b
since $ {(XX^T)}^{-1}(XX^T)$ = Identity matrix as shown above example using Z matrix.

OR b can be written as b = $ {(XX^T)}^{-1} X^TY$ 

In [36]:
mat = X.T.dot(X)

In [37]:
print('X.shape', X.shape)
print('X.T.shape', X.T.shape)
print('y.shape', y.shape)
print('mat.shape', mat.shape)
print('det(mat)', det(mat))

X.shape (50, 4)
X.T.shape (4, 50)
y.shape (50, 1)
mat.shape (4, 4)
det(mat) 244.357283999992


In [38]:
b = inv(mat).dot(X.T).dot(y)

In [39]:
print("\nRegression coefficients\n")
print(b)


Regression coefficients

[[2.34544117]
 [0.65974156]
 [0.25452789]
 [0.13497273]]


**Inference from the solution using linear algebra **

* b0 = 2.3454  (aprroximated to 4 decimals)
* b1 = 0.6597 (aprroximated to 4 decimals)
* b2 = 0.2545  (aprroximated to 4 decimals)
* b3 = 0.1350  (aprroximated to 4 decimals)

y = 2.3454 + (0.6597 * X1) + (0.2545 * X2) + (0.1350 * X3)
* where y represents the sepal length in cm of Setosa species of iris flower
* X1 represents its petal length (cm) and X2 represents petal width (cm)
* 2.3454 is a regression constant
* 0.6597, 0.2545 and 0.1350 are regression coefficients.

#### Practice Exercise: 1

Refer to the above example.

Predict Sepal length from sepal width (cm), petal length (cm), petal width (cm) of the species, Versicolour.
We shall select all rows belonging to class 1, Versicolour.

# END